core/config/options.go

```diff
@@ -0,0 +1,167 @@
+
+var ErrConfigOp = errors.New("invalid config operation")
+
+func New(sdb *sinkdb.DB) *Options {
+	return &Options{
+		sdb:    sdb,
+		schema: make(map[string]option),
+	}
+}
+
+// Options provides access to Chain Core configuration options. All
+// options should be defined before accessing and modifying any
+// of their values.
+type Options struct {
+	sdb    *sinkdb.DB
+	schema map[string]option
+}
+
+// CleanFunc is required when defining a new configuration option.
+// Implementations should canonicalize and validate newTuple. In-place
+// modifications are ok.
+type CleanFunc func(newTuple []string) error
+
+// EqualFunc is required when defining a new set configuration option.
+// Both a and b are guaranteed to have already been cleaned by the
+// option's CleanFunc. It is also guaranteed that len(a) == len(b).
+type EqualFunc func(a, b []string) bool
+
+type option struct {
+	tupleSize int
+	cleanFunc CleanFunc
+	equalFunc EqualFunc
+}
+
+// DefineSet defines a new configuration option as a set of tuples of
+// size tupleSize. New tuples will be validated and canonicalized by
+// cleanFunc. Equality when adding and removing is determined by equalFunc.
+func (opts *Options) DefineSet(key string, tupleSize int, cleanFunc CleanFunc, equalFunc EqualFunc) {
+	opts.schema[key] = option{
+		tupleSize: tupleSize,
+		cleanFunc: cleanFunc,
+		equalFunc: equalFunc,
+	}
+}
+
+// List returns all of the tuples for the provided configuration option.
+func (opts *Options) List(ctx context.Context, key string) ([][]string, error) {
+	if _, ok := opts.schema[key]; !ok {
+		return nil, errors.WithDetailf(ErrConfigOp, "Configuration option %q is undefined.", key)
+	}
+	var set ValueSet
+	_, err := opts.sdb.Get(ctx, filepath.Join(sinkdbPrefix, key), &set)
+	if err != nil {
+		return nil, err
+	}
+	var tuples [][]string
+	for _, tup := range set.Tuples {
+		tuples = append(tuples, tup.Values)
+	}
+	return tuples, nil
+}
+
+// Add adds the provided tuple to the configuration option set indicated
+// by key.
+func (opts *Options) Add(key string, tup []string) (sinkdb.Op, error) {
+	opt, ok := opts.schema[key]
+	if !ok {
+		return sinkdb.Op{}, errors.WithDetailf(ErrConfigOp, "Configuration option %q is undefined.", key)
+	}
+	if opt.tupleSize != len(tup) {
+		return sinkdb.Op{}, errors.WithDetailf(ErrConfigOp, "Configuration option %q expects %d arguments.", key, opt.tupleSize)
+	}
+	if opt.equalFunc == nil {
+		return sinkdb.Op{}, errors.WithDetailf(ErrConfigOp, "Configuration option %q is a scalar. Use corectl set instead.")
+	}
+
+	// make a copy to avoid mutating tup
+	cleaned := make([]string, len(tup))
+	copy(cleaned, tup)
+	err := opt.cleanFunc(cleaned)
+	if err != nil {
+		return sinkdb.Op{}, errors.Sub(ErrConfigOp, err)
+	}
+
+	var existing ValueSet
+	ver, err := opts.sdb.GetStale(filepath.Join(sinkdbPrefix, key), &existing)
+	if err != nil {
+		return sinkdb.Op{}, err
+	}
+	idx := findIndex(existing.Tuples, cleaned, opt.equalFunc)
```

Comment:
Nit: name a pure function after what it returns, rather than what it does (that's for a procedure). So maybe `index` or `tupleIndex`.